File: tests/go/test_client.go
package main

import (
	"fmt"
	"testing"
	"time"
)

// MockClient simulates FastDataBroker client
type MockClient struct {
	brokers []string
	stream  string
}

// Client test structure
type ClientTest struct {
	t      *testing.T
	client *MockClient
}

// NewMockClient creates a test client
func NewMockClient(brokers []string, stream string) *MockClient {
	return &MockClient{
		brokers: brokers,
		stream:  stream,
	}
}

// TestClientInitialization tests client initialization
func TestClientInitialization(t *testing.T) {
	brokers := []string{"broker1:8080", "broker2:8081"}
	client := NewMockClient(brokers, "test-stream")
	
	if client == nil {
		t.Fatal("Client creation failed")
	}
	if client.stream != "test-stream" {
		t.Errorf("Expected stream 'test-stream', got '%s'", client.stream)
	}
	if len(client.brokers) != 2 {
		t.Errorf("Expected 2 brokers, got %d", len(client.brokers))
	}
}

// TestClientConnect simulates broker connection
func TestClientConnect(t *testing.T) {
	client := NewMockClient([]string{"broker1:8080"}, "test")
	
	if len(client.brokers) == 0 {
		t.Fatal("No brokers available")
	}
	
	// Simulate connection
	for _, broker := range client.brokers {
		if len(broker) == 0 {
			t.Errorf("Invalid broker address")
		}
	}
}

// TestDiscoveryTopology tests topology discovery
func TestDiscoveryTopology(t *testing.T) {
	client := NewMockClient([]string{"broker1:8080", "broker2:8081"}, "orders")
	
	// Simulate topology response
	topology := map[string]interface{}{
		"stream":   "orders",
		"brokers":  client.brokers,
		"partition_count": 4,
		"replication_factor": 3,
	}
	
	if topology["stream"] != "orders" {
		t.Errorf("Expected stream 'orders'")
	}
	if topology["partition_count"] != 4 {
		t.Errorf("Expected 4 partitions")
	}
}

// TestProducerSendMessage tests message production
func TestProducerSendMessage(t *testing.T) {
	client := NewMockClient([]string{"broker1:8080"}, "test")
	
	// Simulate send
	message := map[string]interface{}{
		"key": "test-key",
		"value": "test-value",
	}
	
	partition := 0 // hash("test-key") % 4 = 0
	
	if partition < 0 || partition >= 4 {
		t.Errorf("Invalid partition: %d", partition)
	}
	if message["key"] != "test-key" {
		t.Errorf("Message key mismatch")
	}
}

// TestProducerBatch tests batch sending
func TestProducerBatch(t *testing.T) {
	client := NewMockClient([]string{"broker1:8080"}, "test")
	
	// Send 100 messages
	messageCount := 100
	successCount := 0
	
	for i := 0; i < messageCount; i++ {
		// Simulate send success
		successCount++
	}
	
	if successCount != messageCount {
		t.Errorf("Expected %d successes, got %d", messageCount, successCount)
	}
}

// TestConsumerGroupAssignment tests consumer group assignment
func TestConsumerGroupAssignment(t *testing.T) {
	client := NewMockClient([]string{"broker1:8080"}, "test")
	
	// Simulate consumer group with 2 consumers
	groupMembers := []string{"consumer-1", "consumer-2"}
	partitions := []int{0, 1, 2, 3}
	
	// Round-robin assignment
	assignment := make(map[string][]int)
	for i, consumer := range groupMembers {
		assignment[consumer] = []int{partitions[i*len(partitions)/len(groupMembers):
			(i+1)*len(partitions)/len(groupMembers)]}
	}
	
	if len(assignment) != len(groupMembers) {
		t.Errorf("Expected %d consumers assigned", len(groupMembers))
	}
}

// TestConsumerPolling tests consumer message polling
func TestConsumerPolling(t *testing.T) {
	client := NewMockClient([]string{"broker1:8080"}, "test")
	
	// Simulate polling 100 messages
	pollCount := 0
	maxPolls := 100
	timeout := time.Millisecond * 100
	
	for pollCount < maxPolls {
		// Simulate poll
		pollCount++
	}
	
	if pollCount == 0 {
		t.Error("Consumer polling failed")
	}
}

// TestConsumerCommit tests offset commit
func TestConsumerCommit(t *testing.T) {
	client := NewMockClient([]string{"broker1:8080"}, "test")
	
	// Simulate offset commit
	partition := 0
	offset := int64(100)
	
	// Commit offset
	committed := offset
	
	if committed != offset {
		t.Errorf("Offset commit failed: expected %d, got %d", offset, committed)
	}
}

// TestConsistentHashing tests partition selection via consistent hashing
func TestConsistentHashing(t *testing.T) {
	// Test consistent hashing
	keys := []string{"key1", "key2", "key3", "key4"}
	partitions := 4
	
	// Hash function simulation
	hashMap := make(map[string]int)
	for _, key := range keys {
		// Simple hash: sum of ASCII values
		hash := 0
		for _, c := range key {
			hash += int(c)
		}
		partition := hash % partitions
		hashMap[key] = partition
	}
	
	// Verify consistency: same key always maps to same partition
	for _, key := range keys {
		hash := 0
		for _, c := range key {
			hash += int(c)
		}
		partition := hash % partitions
		if hashMap[key] != partition {
			t.Errorf("Hash inconsistency for key %s", key)
		}
	}
}

// TestPartitionDistribution tests load distribution across partitions
func TestPartitionDistribution(t *testing.T) {
	partitionCount := 4
	keyCount := 1000
	distribution := make(map[int]int)
	
	// Distribute 1000 keys across 4 partitions
	for i := 0; i < keyCount; i++ {
		key := fmt.Sprintf("key-%d", i)
		hash := 0
		for _, c := range key {
			hash += int(c)
		}
		partition := hash % partitionCount
		distribution[partition]++
	}
	
	// Check distribution is balanced (within 25%)
	expectedPerPartition := keyCount / partitionCount // 250
	tolerance := expectedPerPartition / 4              // 62 (25%)
	
	for partition, count := range distribution {
		if count < expectedPerPartition-tolerance ||
			count > expectedPerPartition+tolerance {
			t.Logf("Partition %d has %d keys (expected ~%d)", 
				partition, count, expectedPerPartition)
		}
	}
}

// TestReplicationAwareness tests awareness of replica brokers
func TestReplicationAwareness(t *testing.T) {
	client := NewMockClient([]string{"broker1:8080", "broker2:8081", "broker3:8082"}, "test")
	
	// Simulate partition with replicas
	partition := map[string]interface{}{
		"id":      0,
		"leader":  "broker1",
		"replicas": []string{"broker1", "broker2", "broker3"},
		"in_sync_replicas": []string{"broker1", "broker2"},
	}
	
	isr := partition["in_sync_replicas"].([]string)
	if len(isr) < 2 {
		t.Errorf("Expected at least 2 in-sync replicas, got %d", len(isr))
	}
}

// TestFailoverDetection tests broker failure detection
func TestFailoverDetection(t *testing.T) {
	brokers := []string{"broker1:8080", "broker2:8081", "broker3:8082"}
	
	// Simulate broker status
	brokerStatus := map[string]string{
		"broker1": "UP",
		"broker2": "UP",
		"broker3": "DOWN",
	}
	
	// Count healthy brokers
	healthyCount := 0
	for _, status := range brokerStatus {
		if status == "UP" {
			healthyCount++
		}
	}
	
	if healthyCount < 2 {
		t.Errorf("Insufficient healthy brokers: %d", healthyCount)
	}
}

// TestClusterMetrics tests cluster metric collection
func TestClusterMetrics(t *testing.T) {
	client := NewMockClient([]string{"broker1:8080", "broker2:8081"}, "test")
	
	// Simulate metrics
	metrics := map[string]interface{}{
		"broker_count": len(client.brokers),
		"stream_count": 1,
		"partition_count": 4,
		"message_rate": 912000,
		"latency_p99_ms": 2.5,
	}
	
	if metrics["broker_count"].(int) != len(client.brokers) {
		t.Errorf("Broker count mismatch")
	}
	if metrics["message_rate"].(int) != 912000 {
		t.Errorf("Message rate incorrect")
	}
}

// BenchmarkHashFunction benchmarks the hash function
func BenchmarkHashFunction(b *testing.B) {
	keys := []string{"key1", "key2", "key3", "key4"}
	
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		for _, key := range keys {
			hash := 0
			for _, c := range key {
				hash += int(c)
			}
			_ = hash % 4
		}
	}
}

// BenchmarkMessageSend benchmarks message sending
func BenchmarkMessageSend(b *testing.B) {
	client := NewMockClient([]string{"broker1:8080"}, "test")
	
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		// Simulate send
		_ = client
	}
}

func main() {
	fmt.Println("Go SDK tests ready. Run with: go test ./...")
}

File: tests/javascript/test_client.js
/**
 * FastDataBroker JavaScript SDK Test Suite
 */

const assert = require('assert');

/**
 * Mock Client for testing
 */
class MockClient {
    constructor(brokers, streamName) {
        this.brokers = brokers;
        this.streamName = streamName;
        this.partitionCount = 4;
    }
    
    getStreamName() {
        return this.streamName;
    }
    
    getBrokerCount() {
        return this.brokers.length;
    }
    
    getBrokers() {
        return [...this.brokers];
    }
    
    getTopology() {
        return {
            stream: this.streamName,
            broker_count: this.brokers.length,
            partition_count: this.partitionCount,
            replication_factor: 3
        };
    }
}

/**
 * Producer for testing
 */
class Producer {
    constructor(client) {
        this.client = client;
    }
    
    send(key, value) {
        const hash = this.hashKey(key);
        return hash % 4;
    }
    
    hashKey(key) {
        let hash = 0;
        for (let i = 0; i < key.length; i++) {
            hash += key.charCodeAt(i);
        }
        return Math.abs(hash);
    }
}

/**
 * Consumer for testing
 */
class Consumer {
    constructor(client, groupId) {
        this.client = client;
        this.groupId = groupId;
        this.offsets = {};
    }
    
    commitOffset(partition, offset) {
        this.offsets[partition] = offset;
    }
    
    getCommittedOffset(partition) {
        return this.offsets[partition] || 0;
    }
    
    async poll(timeout = 5000) {
        return new Promise((resolve) => {
            setTimeout(() => {
                resolve(null);
            }, timeout);
        });
    }
}

/**
 * Partition for testing
 */
class Partition {
    constructor(id, replicas) {
        this.id = id;
        this.replicas = replicas;
        this.leader = replicas[0];
        this.inSyncReplicas = replicas.slice(0, 2);
    }
}

/**
 * Test Suite
 */
describe('FastDataBroker JavaScript SDK', () => {
    let client;
    
    beforeEach(() => {
        const brokers = [
            'broker1:8080',
            'broker2:8081',
            'broker3:8082',
            'broker4:8083'
        ];
        client = new MockClient(brokers, 'test-stream');
    });
    
    describe('Client Initialization', () => {
        it('should create client with brokers', () => {
            assert(client !== null);
            assert.equal(client.getStreamName(), 'test-stream');
            assert.equal(client.getBrokerCount(), 4);
        });
        
        it('should have valid broker list', () => {
            const brokers = client.getBrokers();
            assert.equal(brokers.length, 4);
            assert(brokers[0].includes(':'));
        });
    });
    
    describe('Topology Discovery', () => {
        it('should retrieve cluster topology', () => {
            const topology = client.getTopology();
            assert.equal(topology.stream, 'test-stream');
            assert.equal(topology.broker_count, 4);
            assert.equal(topology.partition_count, 4);
            assert.equal(topology.replication_factor, 3);
        });
    });
    
    describe('Producer', () => {
        let producer;
        
        beforeEach(() => {
            producer = new Producer(client);
        });
        
        it('should send message to partition', () => {
            const partition = producer.send('test-key', 'test-value');
            assert(partition >= 0 && partition < 4);
        });
        
        it('should send multiple messages', () => {
            for (let i = 0; i < 100; i++) {
                const partition = producer.send(`key-${i}`, `value-${i}`);
                assert(partition >= 0 && partition < 4);
            }
        });
        
        it('should hash same key to same partition', () => {
            const partition1 = producer.send('same-key', 'value1');
            const partition2 = producer.send('same-key', 'value2');
            assert.equal(partition1, partition2);
        });
        
        it('should distribute keys evenly', () => {
            const distribution = [0, 0, 0, 0];
            const keyCount = 1000;
            
            for (let i = 0; i < keyCount; i++) {
                const partition = producer.send(`key-${i}`, `value-${i}`);
                distribution[partition]++;
            }
            
            // Check balance (within 25%)
            const expectedPerPartition = keyCount / 4; // 250
            const tolerance = expectedPerPartition / 4; // 62
            
            distribution.forEach((count, partition) => {
                assert(
                    count >= expectedPerPartition - tolerance &&
                    count <= expectedPerPartition + tolerance,
                    `Partition ${partition} has ${count} keys (expected ~${expectedPerPartition})`
                );
            });
        });
    });
    
    describe('Consumer', () => {
        let consumer;
        
        beforeEach(() => {
            consumer = new Consumer(client, 'test-group');
        });
        
        it('should commit offset', () => {
            consumer.commitOffset(0, 100);
            assert.equal(consumer.getCommittedOffset(0), 100);
        });
        
        it('should handle multiple partitions', () => {
            for (let i = 0; i < 4; i++) {
                consumer.commitOffset(i, i * 100);
            }
            
            for (let i = 0; i < 4; i++) {
                assert.equal(consumer.getCommittedOffset(i), i * 100);
            }
        });
        
        it('should return 0 for uncommitted partition', () => {
            assert.equal(consumer.getCommittedOffset(99), 0);
        });
        
        it('should poll messages with timeout', async () => {
            const message = await consumer.poll(100);
            // Message will be null in mock
            assert(message === null);
        }).timeout(500);
    });
    
    describe('Partitioning', () => {
        it('should create partition with replicas', () => {
            const replicas = ['broker0', 'broker1', 'broker2'];
            const partition = new Partition(0, replicas);
            
            assert.equal(partition.id, 0);
            assert.equal(partition.leader, 'broker0');
            assert.equal(partition.inSyncReplicas.length, 2);
        });
        
        it('should handle partition failover', () => {
            const partition = new Partition(0, ['broker0', 'broker1', 'broker2']);
            
            // Simulate broker0 failure
            partition.inSyncReplicas = ['broker1', 'broker2'];
            partition.leader = 'broker1'; // Elect new leader
            
            assert.equal(partition.leader, 'broker1');
            assert.equal(partition.inSyncReplicas.length, 2);
        });
    });
    
    describe('Replication', () => {
        it('should have 3-way replication', () => {
            const partition = new Partition(0, ['broker0', 'broker1', 'broker2']);
            assert.equal(partition.replicas.length, 3);
        });
        
        it('should tolerate 1 broker failure', () => {
            let replicas = ['broker0', 'broker1', 'broker2'];
            
            // Simulate 1 failure
            replicas = replicas.filter(b => b !== 'broker0');
            
            assert.equal(replicas.length, 2);
            assert(replicas.includes('broker1'));
            assert(replicas.includes('broker2'));
        });
    });
    
    describe('Hash Performance', () => {
        it('should hash keys quickly', () => {
            const producer = new Producer(client);
            const iterations = 10000;
            
            const startTime = Date.now();
            for (let i = 0; i < iterations; i++) {
                producer.send(`key-${i}`, `value-${i}`);
            }
            const endTime = Date.now();
            
            const durationMs = endTime - startTime;
            console.log(`Hashed ${iterations} keys in ${durationMs}ms`);
            assert(durationMs < 1000, 'Hashing should be fast');
        });
    });
    
    describe('Concurrent Operations', () => {
        it('should handle concurrent sends', async () => {
            const producer = new Producer(client);
            const promises = [];
            
            for (let i = 0; i < 100; i++) {
                promises.push(
                    Promise.resolve(producer.send(`key-${i}`, `value-${i}`))
                );
            }
            
            const results = await Promise.all(promises);
            assert.equal(results.length, 100);
        });
    });
    
    describe('Error Handling', () => {
        it('should handle invalid partition gracefully', () => {
            const producer = new Producer(client);
            const partition = producer.send('test', 'value');
            
            assert(partition >= 0 && partition < 4, 'Partition should be valid');
        });
    });
});

/**
 * Run tests with: npm test
 * Or with mocha: mocha test_client.js
 */

// Export for use in other tests
module.exports = {
    MockClient,
    Producer,
    Consumer,
    Partition
};
